File: app/src/main/java/com/denchic45/kts/data/model/mapper/CourseContentMapper.java
package com.denchic45.kts.data.model.mapper;

import androidx.annotation.NonNull;

import com.denchic45.kts.data.model.domain.Attachment;
import com.denchic45.kts.data.model.domain.ContentDetails;
import com.denchic45.kts.data.model.domain.ContentType;
import com.denchic45.kts.data.model.domain.CourseContent;
import com.denchic45.kts.data.model.domain.Task;
import com.denchic45.kts.data.model.firestore.CourseContentDoc;
import com.denchic45.kts.data.model.room.CourseContentEntity;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import kotlin.Pair;

@Mapper
public interface CourseContentMapper {

    @Mapping(source = "courseEntity.id", target = "id")
    Task courseContentEntityWithDetailsToTask(CourseContentEntity courseEntity, ContentDetails.Task taskDetails);

    @Mapping(source = "task.id", target = "id")
    @Mapping(target = "submissions", ignore = true)
    CourseContentDoc taskWithDetailsToTaskDoc(Task task, ContentDetails.Task contentDetails, ContentType contentType);

    default Pair<List<CourseContentEntity>, List<CourseContentDoc>> docToEntity(@NonNull QuerySnapshot snapshots) {
        List<CourseContentEntity> courseContentEntities = new ArrayList<>();
        List<CourseContentDoc> courseContentDocs = new ArrayList<>();
        for (QueryDocumentSnapshot snapshot : snapshots) {
            CourseContentDoc courseContentDoc = snapshot.toObject(CourseContentDoc.class);
            courseContentEntities.add(
                    courseContentDocToEntity(
                            courseContentDoc,
                            new Gson().toJson(snapshot.get("details"))
                    )
            );
            courseContentDocs.add(courseContentDoc);
        }
        return new Pair<>(courseContentEntities, courseContentDocs);
    }

    @Mapping(source = "details", target = "details")
    CourseContentEntity courseContentDocToEntity(CourseContentDoc courseContentDoc, String details);

    default CourseContentDoc domainToDoc(CourseContent courseContent) {
        if (courseContent instanceof Task) {
            return domainToTaskDoc((Task) courseContent);
        } else
            throw new IllegalStateException();
    }

    default CourseContentDoc domainToTaskDoc(Task task) {
        return taskWithDetailsToTaskDoc(task, taskToTaskDetails(task), ContentType.TASK);
    }

    ContentDetails.Task taskToTaskDetails(Task task);

    default Task entityToTaskDomain(CourseContentEntity entity) {
        return courseContentEntityWithDetailsToTask(entity,
                new GsonBuilder().create().fromJson(entity.getDetails(), ContentDetails.Task.class));
    }

    default List<String> mapAttachmentsToFilePaths(@NonNull List<Attachment> attachments) {
        return attachments.stream()
                .map(attachment -> attachment.getFile().getPath())
                .collect(Collectors.toList());
    }

    default List<Attachment> mapFilePathsToAttachments(List<String> filePaths) {
        if (filePaths == null)
            return Collections.emptyList();
        return filePaths.stream()
                .map(path -> new Attachment(new File(path)))
                .collect(Collectors.toList());
    }

    @Named("entityToDomain")
    default CourseContent entityToDomain(@NonNull CourseContentEntity entity) {
        switch (entity.getContentType()) {
            case TASK:
                return entityToTaskDomain(entity);
            default:
                throw new IllegalStateException();
        }
    }

    List<CourseContent> entityToDomain(List<CourseContentEntity> courseContentEntities);
}

File: app/src/main/java/com/denchic45/kts/data/model/mapper/CourseContentMapper.kt
package com.denchic45.kts.data.model.mapper

import com.denchic45.kts.data.domain.model.Attachment
import com.denchic45.kts.data.model.room.CourseContentEntity
import com.denchic45.kts.data.remote.model.CourseContentDoc
import com.denchic45.kts.domain.model.ContentDetails
import com.denchic45.kts.domain.model.ContentType
import com.denchic45.kts.domain.model.CourseContent
import com.denchic45.kts.domain.model.Task
import com.google.firebase.firestore.QuerySnapshot
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.Named
import java.io.File

@Mapper
abstract class CourseContentMapper {
    @Mapping(source = "courseEntity.id", target = "id")
    abstract fun courseContentEntityWithDetailsToTask(
        courseEntity: CourseContentEntity,
        taskDetails: ContentDetails.Task,
    ): Task

    @Mapping(source = "courseEntity.id", target = "id")
    @Mapping(source = "contentDetails", target = "contentDetails")
    abstract fun courseContentEntityWithDetailsToTaskDoc(
        courseEntity: CourseContentEntity,
        contentDetails: ContentDetails.Task,
    ): CourseContentDoc

    abstract fun taskToTaskDetails(task: Task): ContentDetails.Task

    @Mapping(source = "task.id", target = "id")
    @Mapping(target = "submissions", ignore = true)
    abstract fun taskWithDetailsToTaskDoc(
        task: Task,
        contentDetails: ContentDetails.Task,
        contentType: ContentType,
    ): CourseContentDoc

    fun docToEntity(snapshots: QuerySnapshot): Pair<List<CourseContentEntity>, List<CourseContentDoc>> {
        val courseContentEntities: MutableList<CourseContentEntity> = ArrayList()
        val courseContentDocs: MutableList<CourseContentDoc> = ArrayList()
        for (snapshot in snapshots) {
            val courseContentMap = snapshot.toObject(CourseContentDoc::class.java)
            courseContentEntities.add(
                courseContentDocToEntity(
                    courseContentMap,
                    Gson().toJson(snapshot["contentDetails"])
                )
            )
            courseContentDocs.add(courseContentMap)
        }
        return Pair<List<CourseContentEntity>, List<CourseContentDoc>>(
            courseContentEntities,
            courseContentDocs
        )
    }

    @Mapping(source = "contentDetails", target = "contentDetails")
    abstract fun courseContentDocToEntity(
        courseContentDoc: CourseContentDoc,
        contentDetails: String,
    ): CourseContentEntity

    fun domainToDoc(courseContent: CourseContent): CourseContentDoc {
        return if (courseContent is Task) {
            domainToTaskDoc(courseContent)
        } else throw IllegalStateException()
    }

    fun domainToTaskDoc(task: Task): CourseContentDoc {
        return taskWithDetailsToTaskDoc(
            task,
            taskToTaskDetails(task),
            ContentType.TASK
        )
    }


    fun entityToTaskDomain(entity: CourseContentEntity): Task {
        return courseContentEntityWithDetailsToTask(
            entity,
            GsonBuilder().create().fromJson(entity.contentDetails, ContentDetails.Task::class.java)
        )
    }

    fun entityToTaskDoc(entity: CourseContentEntity): CourseContentDoc {
        return courseContentEntityWithDetailsToTaskDoc(
            entity,
            GsonBuilder().create().fromJson(entity.contentDetails, ContentDetails.Task::class.java)
        )
    }

    fun mapAttachmentsToFilePaths(attachments: List<Attachment>): List<String> {
        return attachments.map { (file) -> file.path }
    }

    fun mapFilePathsToAttachments(filePaths: List<String>?): List<Attachment> {
        return filePaths?.map { path: String -> Attachment(File(path)) } ?: emptyList()
    }

    @Named("entityToDomain")
    fun entityToDomain(entity: CourseContentEntity): CourseContent {
        return when (entity.contentType) {
            ContentType.TASK -> entityToTaskDomain(entity)
            else -> throw IllegalStateException()
        }
    }

    @Named("entityToDomainAssignment")
    fun entityToDomainAssignment(entity: CourseContentEntity): Task {
        return when (entity.contentType) {
            ContentType.TASK -> entityToTaskDomain(entity)
            else -> throw IllegalStateException()
        }
    }

    abstract fun entityToDomain(courseContentEntities: List<CourseContentEntity>): List<Task>

    abstract fun entityToDomainAssignment(courseContentEntities: List<CourseContentEntity>): List<Task>
}